### PR TITLE
Update FinalizeCredential, use in add-credential

### PR DESCRIPTION
**Pull request by axw**
This diff updates the FinalizeCredential provider
method to take cloud endpoints, so that providers
can communicate with the cloud in order to finalize
the credential. Specifically, Azure needs to connect
to Active Directory to perform interactive authentication,
and then configure the service principal.

**QA**

With a follow-up branch that adds interactive auth-type
for azure, I have done the following:

 1. juju add-credential azure
 2. specify credential name, and select "interactive" auth-type
 3. go to microsoft login page as directed, wait while Juju
    sets up the service principal
 4. bootstrap with the new credential
 5. bob's your uncle

It is also possible, even if undesirable, to add an
interactive auth-type credential in credentials.yaml by
hand, and go through the above steps at bootstrap or
add-model time. The service principal password will not
be persisted to disk then.

### CODE REVIEW

File: cmd/juju/controller/listcontrollersformatters.go [hunk 1]
```diff
@@ -35,7 +35,8 @@ func formatControllersTabular(writer io.Writer, set ControllerSet, promptRefresh
 	w := output.Wrapper{tw}
 
 	if promptRefresh && len(set.Controllers) > 0 {
-		fmt.Fprintln(writer, "Use --refresh to see the latest information.\n")
+		fmt.Fprintln(writer, "Use --refresh to see the latest information.")
+		fmt.Fprintln(writer)
```
Comment (anastasiamac) [inline cmd/juju/controller/listcontrollersformatters.go]: Is this a drive-by?

File: cmd/modelcmd/credentials.go [hunk 2]
```diff
@@ -21,26 +21,60 @@ var (
...
+type GetCredentialsParams struct {
+	// Cloud is the cloud definition.
+	Cloud cloud.Cloud
+
+	// CloudName is the name of the cloud for which credentials are being
+	// obtained.
+	CloudName string
+
+	// CloudRegion is the name of the region that the user has specified.
+	// If this is empty, then GetCredentials will determine the default
```
Comment (anastasiamac) [inline cmd/modelcmd/credentials.go]: Could we please be a bit more precise of what "default" means here? Maybe even say that it's the first region in a list that is sorted in alphabetical order? :D

File: cmd/juju/cloud/addcredential_test.go [hunk 3]
```diff
@@ -207,6 +209,36 @@ func (s *addCredentialSuite) TestAddCredentialMultipleAuthType(c *gc.C) {
...
+Enter username: generating userpass credential
+`[1:])
+
+	// FinalizeCredential should have generated a userpass credential
+	// based on the input from the interactive credential.
+	c.Assert(s.store.Credentials, jc.DeepEquals, map[string]jujucloud.CloudCredential{
+		"somecloud": {
+			AuthCredentials: map[string]jujucloud.Credential{
+				"fred": jujucloud.NewCredential(jujucloud.UserPassAuthType, map[string]string{
+					"username":             "user",
```
Comment (anastasiamac) [inline cmd/juju/cloud/addcredential_test.go]: I know it's a test and this is test data but username = "user" caught me off-guard \o/
Could this be "username" too for clarity?

File: cmd/modelcmd/credentials.go [hunk 4]
```diff
@@ -21,26 +21,60 @@ var (
...
+		if regionName == "" && len(args.Cloud.Regions) > 0 {
+			// No region was specified, use the first region
+			// in the list.
+			regionName = args.Cloud.Regions[0].Name
+		}
+	}
+
+	cloudEndpoint := args.Cloud.Endpoint
+	cloudIdentityEndpoint := args.Cloud.IdentityEndpoint
+	if regionName != "" {
```
Comment (anastasiamac) [inline cmd/modelcmd/credentials.go]: From the reading of above code, the possibility that regionName == "" is slim. Either we do not need this check or we need to do something for the case if it is "". No?

Comment (anastasiamac) [inline cmd/modelcmd/credentials.go]: We should definitely have tests for each of these if- conditions.

File: cmd/juju/controller/listcontrollersformatters.go (hunk 1 above)
Comment (axw) [inline cmd/juju/controller/listcontrollersformatters.go]: Yes, to fix a govet warning.

File: cmd/modelcmd/credentials.go (hunk 2 above)
Comment (axw) [inline cmd/modelcmd/credentials.go]: Expanded comment.

File: cmd/modelcmd/credentials.go (hunk 4 above)
Comment (axw) [inline cmd/modelcmd/credentials.go]: regionName will be "" if there are no regions in the cloud. I've added more tests.

File: cmd/juju/cloud/addcredential_test.go (hunk 3 above)
Comment (axw) [inline cmd/juju/cloud/addcredential_test.go]: Done.

**Comment by axw** [comment]
$$merge$$

**Comment by axw** [comment]
$$merge$$

**Comment by axw** [comment]
$$merge$$
